**Titanic dataset analysis**

This is my analysis of the dataset provided in the kaggle platform regarding one of the most 
infamous shipwrecks in history,'The Titanic'.The follwoing is the link to the dataset :
"https://www.kaggle.com/datasets/yasserh/titanic-dataset"

In [67]:
import pandas as pd

In [68]:
data = pd.read_csv('Titanic-Dataset.csv')

In [85]:
df = pd.DataFrame(data)

In [69]:
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

1.	What is the size of the dataset (number of rows and columns)?

In [70]:
data.shape

(891, 12)

So the number of rows is 891 and columns is 12

2.	Define the columns in dataset

In [71]:
data.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

3.	View last 10 rows

In [73]:
data.tail(10)

PassengerId  Survived  Pclass                                      Name  \
881          882         0       3                        Markun, Mr. Johann   
882          883         0       3              Dahlberg, Miss. Gerda Ulrika   
883          884         0       2             Banfield, Mr. Frederick James   
884          885         0       3                    Sutehall, Mr. Henry Jr   
885          886         0       3      Rice, Mrs. William (Margaret Norton)   
886          887         0       2                     Montvila, Rev. Juozas   
887          888         1       1              Graham, Miss. Margaret Edith   
888          889         0       3  Johnston, Miss. Catherine Helen "Carrie"   
889          890         1       1                     Behr, Mr. Karl Howell   
890          891         0       3                       Dooley, Mr. Patrick   

        Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  
881    male  33.0      0      0            349257   7.8958   NaN        S  
882  female  22.0      0      0              7552  10.5167   NaN        S  
883    male  28.0      0      0  C.A./SOTON 34068  10.5000   NaN        S  
884    male  25.0      0      0   SOTON/OQ 392076   7.0500   NaN        S  
885  female  39.0      0      5            382652  29.1250   NaN        Q  
886    male  27.0      0      0            211536  13.0000   NaN        S  
887  female  19.0      0      0            112053  30.0000   B42        S  
888  female   NaN      1      2        W./C. 6607  23.4500   NaN        S  
889    male  26.0      0      0            111369  30.0000  C148        C  
890    male  32.0      0      0            370376   7.7500   NaN        Q

4.	How many missing values are there in each column?

In [74]:
data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

There are 177 missing values in 'Age' column, 687 missing values in 'Cabin' column and 2 missing values in 'Embarked' column

5.	What is the distribution of passenger ages in the dataset?

In [13]:
data.groupby('PassengerId')['Age'].mean()

PassengerId
1      22.0
2      38.0
3      26.0
4      35.0
5      35.0
       ... 
887    27.0
888    19.0
889     NaN
890    26.0
891    32.0
Name: Age, Length: 891, dtype: float64

6.	How many unique values are present in the 'Embarked' column?

In [21]:
data['Embarked'].nunique()

3

7.	What is the most common surname among passengers?

In [76]:
def surname(name):
    if ',' in name:
        return name.split(",")[0].strip()
    else:
        return name.split()[0].strip()

data['Surname'] = data['Name'].apply(surname)
surname_counts = data.groupby('Surname').size().reset_index(name='Count')
surname_counts = surname_counts.sort_values(by='Count', ascending=False)
print(surname_counts.iloc[0])

Surname    Andersson
Count              9
Name: 16, dtype: object


The most common surname among passengers is 'Andersson' with 9 people having the surname.

8.	Are there any duplicate records in the dataset?

In [108]:
data.duplicated().sum()

0

So the answer is No, there are no duplicate records in the dataset.

9.  What is the distribution of ticket classes (Pclass) among passengers?

In [77]:
data['Pclass'].value_counts()

Pclass
3    491
1    216
2    184
Name: count, dtype: int64

10.	How many passengers have missing values for 'Cabin' information?

In [78]:
data['Cabin'].isnull().sum()

687

687 passengers have missing values for 'Cabin' information

11.	What is the average number of siblings/spouses (SibSp) aboard the Titanic?

In [79]:
 data['SibSp'].mean()

0.5230078563411896

The average number of siblings/spouses aboard the Titanic is 0.52

12.	What is the proportion of male versus female passengers?

In [107]:
data['Sex'].value_counts(normalize=True)*100

Sex
male      64.758698
female    35.241302
Name: proportion, dtype: float64

The proportion of male passengers is 0.64 and of female passengers is 0.35

13.	How many passengers paid a fare of zero?

In [102]:
df[df['Fare'] == 0].shape[0]

15

15 passengers paid a fare of zero

14.	Perform feature engineering to create a new feature 'Title' from the 'Name' column and analyze the survival rate based on this new feature.

In [104]:
df['Title'] = df['Name'].str.extract(' ([A-Za-z]+)\.', expand=False)
print(df['Title'].unique())
title_survival_rate = df.groupby('Title')['Survived'].mean().sort_values(ascending=False)
print(title_survival_rate)

['Mr' 'Mrs' 'Miss' 'Master' 'Don' 'Rev' 'Dr' 'Mme' 'Ms' 'Major' 'Lady'
 'Sir' 'Mlle' 'Col' 'Capt' 'Countess' 'Jonkheer']
Title
Sir         1.000000
Countess    1.000000
Ms          1.000000
Mme         1.000000
Lady        1.000000
Mlle        1.000000
Mrs         0.792000
Miss        0.697802
Master      0.575000
Col         0.500000
Major       0.500000
Dr          0.428571
Mr          0.156673
Jonkheer    0.000000
Don         0.000000
Rev         0.000000
Capt        0.000000
Name: Survived, dtype: float64


This is the survival rate from most to least according to 'Title' of the passengers

15.	Group the passengers by 'Pclass' and calculate the average 'Fare' and 'Age' for each class.

In [ ]:
#Method 1

In [37]:
data.groupby('Pclass')[['Fare','Age']].mean()

Fare        Age
Pclass                      
1       84.154687  38.233441
2       20.662183  29.877630
3       13.675550  25.140620

In [ ]:
#Method 2

In [113]:
data.groupby('Pclass').agg(Fare=('Fare','mean') , Age=('Age','mean'))

Fare        Age
Pclass                      
1       84.154687  38.233441
2       20.662183  29.877630
3       13.675550  25.140620

16.	Group the passengers by 'Sex' and 'Pclass' and find the median 'Age' for each group.

In [39]:
data.groupby(['Sex','Pclass'])['Age'].median()

Sex     Pclass
female  1         35.0
        2         28.0
        3         21.5
male    1         40.0
        2         30.0
        3         25.0
Name: Age, dtype: float64

17.	Create a new dataframe that shows the count of passengers and the average 'Fare' for each combination of 'Pclass' and 'Embarked'.

In [98]:
grouped_df = df.groupby(['Pclass', 'Embarked']).agg(Passenger_Count=('PassengerId', 'size'),Average_Fare=('Fare', 'mean')).reset_index()
grouped_df = grouped_df.set_index('Passenger_Count').reset_index()
grouped_df

Passenger_Count  Pclass Embarked  Average_Fare
0               85       1        C    104.718529
1                2       1        Q     90.000000
2              127       1        S     70.364862
3               17       2        C     25.358335
4                3       2        Q     12.350000
5              164       2        S     20.327439
6               66       3        C     11.214083
7               72       3        Q     11.183393
8              353       3        S     14.644083